pep-0684.rst

```diff
@@ -0,0 +1,640 @@
+   (including objects) into ``_PyRuntimeState``
+2. move nearly all of the state down into ``PyInterpreterState``
+3. finally, move the GIL down into ``PyInterpreterState``
+4. everything else
+
+   * add to the public C-API
+   * implement restrictions in ``ExtensionFileLoader``
+   * work with popular extension maintainers to help with multi-interpreter support
+
+Per-Interpreter State
+---------------------
+
+The following runtime state will be moved to ``PyInterpreterState``:
+
+* all global objects that are not safely shareable (fully immutable)
+* the GIL
+* mutable, currently protected by the GIL
+* mutable, currently protected by some other per-interpreter lock
+* mutable, may be used independently in different interpreters
+* all other mutable (or effectively mutable) state not otherwise excluded below
+
+The following state will not be moved:
+
+* global objects that are safely shareable, if any
+* immutable, often ``const``
+* treated as immutable
+* related to CPython's ``main()`` execution
+* related to the REPL
+* set during runtime init, then treated as immutable
+* mutable, protected by some global lock
+* mutable, atomic
+
+Also, a number of parts of the global state were already moved to the
+interpreter, such as GC, warnings, and atexit hooks.
+
+Note that currently the allocators (see Objects/obmalloc.c) are shared
+between all interpreters, protected by the GIL.  They will need to move
+to each interpreter (or a global lock will be needed).  This is the
+highest risk part of the work to isolate interpreters and may require
+more than just moving fields down from ``_PyRuntimeState``.  Some of
+the complexity is reduced if CPython switches to a thread-safe
+allocator like mimalloc.
+
+.. _proposed capi:
+
+C-API
+-----
+
+The following private API will be made public:
+
+* ``_PyInterpreterConfig``
+* ``_Py_NewInterpreter()`` (as ``Py_NewInterpreterEx()``)
+
+The following fields will be added to ``PyInterpreterConfig``:
+
+* ``own_gil`` - (bool) create a new interpreter lock
+  (instead of sharing with the main interpreter)
+* ``strict_extensions`` - fail import in this interpreter for
+  incompatible extensions (see `Restricting Extension Modules`_)
+
+Restricting Extension Modules
+-----------------------------
+
+Extension modules have many of the same problems as the runtime when
+state is stored in global variables.  :pep:`630` covers all the details
+of what extensions must do to support isolation, and thus safely run in
+multiple interpreters at once.  This includes dealing with their globals.
+
+Extension modules that do not implement isolation will only run in
+the main interpreter.  In all other interpreters the import will
+raise ``ImportError``.  This will be done through
+``importlib._bootstrap_external.ExtensionFileLoader``.
+
+We will work with popular extensions to help them support use in
+multiple interpreters.  This may involve adding new public APIs, which
+we will address on a case-by-case basis.
```

Comment:
```suggestion
multiple interpreters.  This may involve adding new public APIs
in CPython, which we will address on a case-by-case basis.
```

Clarify that you mean public APIs in _CPython_ rather than the extension module (right? I was confused on this bit.)

Reply:
```suggestion
multiple interpreters.  This may involve adding to CPython's public C-API,
which we will address on a case-by-case basis.
```